src/pool/srv_target.c

```diff
@@ -496,14 +510,65 @@ ds_pool_put(struct ds_pool *pool)
 	ABT_mutex_unlock(pool_cache_lock);
 }
 
+#define MAX_HDL_BUF_SIZE	1024
+void
+pool_fetch_hdls_ult(void *data)
+{
+	struct ds_pool	*pool = data;
+	char		buf[MAX_HDL_BUF_SIZE];
+	d_iov_t		iov = { 0 };
+	int		rc = 0;
+
+	ABT_mutex_lock(pool->sp_mutex);
+	if (pool->sp_map == NULL)
+		ABT_cond_wait(pool->sp_fetch_hdls_cond, pool->sp_mutex);
+	ABT_mutex_unlock(pool->sp_mutex);
+retry:
+	if (pool->sp_stopping) {
+		D_DEBUG(DB_MD, DF_UUID" skip fetching hdl due to stop\n",
+			DP_UUID(pool->sp_uuid));
+		D_GOTO(out, rc);
+	}
+	d_iov_set(&iov, buf, MAX_HDL_BUF_SIZE);
+	rc = ds_pool_iv_conn_hdl_fetch(pool, NULL, &iov);
+	if (rc) {
+		if (rc == -DER_REC2BIG) {
+			D_ASSERT(iov.iov_len > MAX_HDL_BUF_SIZE);
+			D_ALLOC(iov.iov_buf, iov.iov_len);
+			if (iov.iov_buf == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			iov.iov_buf_len = MAX_HDL_BUF_SIZE;
+			D_DEBUG(DB_MD, "realloc "DF_U64" and retry.\n",
+				iov.iov_len);
+			D_GOTO(retry, rc);
```

Comment:
The code should be definite here, it can only retry once.

Reply:
well, it will reallocate the buffer by the return (specified) size, so it should retry more until the buffer size satisfied the needs.

Reply:
Seems odd to me to loop with a GOTO where a while loop would work... Just another unfortunate side effect of the 80 character limit I guess.

Reply:
> well, it will reallocate the buffer by the return (specified) size, so it should retry more until the buffer size satisfied the needs.

I don't see why it can still hit REC2GIG after allocating sufficient buffer. If you want multiple retries, you need to guarantee no memory leak and no infinite retry.

Reply:
If others keeps connect  the pool, the pool iv connect buffer will keep increasing.  then it might keeps hitting REC2BIG .